spring-cloud-stream/src/test/java/org/springframework/cloud/stream/function/FunctionSupportTests.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.scheduling.support.PeriodicTrigger;
+import org.springframework.util.MimeTypeUtils;
+import org.springframework.util.StringUtils;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ *
+ * @author Oleg Zhurakousky
+ *
+ * @since 2.1
+ *
+ */
+public class FunctionSupportTests {
+
+	@Test
+	public void testFunctionIsAppliedToMessageSource() {
+		ApplicationContext context = new SpringApplicationBuilder(TestChannelBinderConfiguration.getCompleteConfiguration(MyFunctionsConfiguration.class))
+				.web(WebApplicationType.NONE).run("--spring.cloud.stream.function.name=toUpperCase",  "--spring.jmx.enabled=false");
+
+		OutputDestination target = context.getBean(OutputDestination.class);
+		assertThat(target.receive(1000).getPayload()).isEqualTo("HELLO FUNCTION".getBytes(StandardCharsets.UTF_8));
+	}
+
+	@Test
+	public void testComposedFunctionIsAppliedToMessageSource() {
+		ApplicationContext context = new SpringApplicationBuilder(TestChannelBinderConfiguration.getCompleteConfiguration(MyFunctionsConfiguration.class))
+				.web(WebApplicationType.NONE).run("--spring.cloud.stream.function.name=toUpperCase|concatWithSelf", "--spring.jmx.enabled=false");
+
+		OutputDestination target = context.getBean(OutputDestination.class);
+		assertThat(target.receive(1000).getPayload()).isEqualTo("HELLO FUNCTION:HELLO FUNCTION".getBytes(StandardCharsets.UTF_8));
+	}
+
+	@SpringBootApplication
+	@Import(SourceConfiguration.class)
+	public static class MyFunctionsConfiguration {
+		@Bean
+		public Function<String, String> toUpperCase() {
+			return x -> x.toUpperCase();
+		}
+
+		@Bean
+		public Function<String, String> concatWithSelf() {
+			return x -> x + ":" + x;
+		}
+	}
+
+	/**
+	 * This configuration essentially emulates our existing app-startes for Sources
+	 * and essentially demonstrates how a function(s) could be applied to an existing
+	 * source via {@link FunctionSupport} class.
+	 */
+	@EnableBinding(Source.class)
+	public static class SourceConfiguration {
+
+		@Autowired
+		private Source source;
+
+		@Autowired
+		private FunctionProperties functionProperties;
+
+		@Bean
+		public IntegrationFlow messageSourceFlow(FunctionSupport functionSupport) {
+			Supplier<Message<String>> messageSource = () -> MessageBuilder.withPayload("hello function")
+					.setHeader(MessageHeaders.CONTENT_TYPE, MimeTypeUtils.TEXT_PLAIN).build();
+
+			PollerMetadata pm = new PollerMetadata();
+			pm.setTrigger(new PeriodicTrigger(1, TimeUnit.SECONDS));
+			IntegrationFlowBuilder flowBuilder = IntegrationFlows.from(messageSource);
+
+			if (StringUtils.hasText(functionProperties.getName())) {
+				functionSupport.applyFunctionToIntegrationFlow(functionProperties.getName(), flowBuilder, message -> source.output().send(message));
```

Comment:
As I said before this is a bit flaky and I don't think that it is somehow related to the non-SI Binder implementation.
This is just a matter of something like `.fluxTransform(Function<Flux<?>, Flux<?>>)` in the Spring Integration `IntegrationFlowDefinition`. I mean you definitely have something like this in SCSt via `@StreamListener` with the `Flux` input/output, so, I don't see a big deal to introduce something like that for SI as well, especially when we really have there a `FluxMessageChannel` already.

Yeah... This might need a separate story and we may really live with your solution for a while... Although this is a public API, end-user matter 😕

Reply:
The idea is for FunctionInvoker to evolve and handle both SI and non-SI binders, so providing that extra consumer is still required. In fact FunctionInvoker in this PR is identical to the one in non-SI binder branch with the except with less features (no windowing, reties etc). So we're good here, but yes, feel free to start a story under #1436 so we can track all your thoughts/ideas and bring it up to reactor team.